optpresso/server/static/js/app.js

```diff
@@ -0,0 +1,394 @@
+    }
+    mean = Number((mean).toFixed(2));
+    std = Number((std).toFixed(2));
+    return (
+      <div class="container">
+        <div class="row">
+          <div class="col-md-5 offset-1 config">
+            <h3>Predictions</h3>
+            Mean:{mean}<br></br>
+            Std: {std}
+            <ul>{listItems}</ul>
+            <button onClick={this.clearPredictions}>Clear Predictions</button>
+          </div>
+          <div class="col-md-5 offset-1 config">
+            <button onClick={this.handlePredict}>Predict</button>
+          </div>
+        </div>
+      </div>
+      
+    )
+  }
+}
+
+class CaptureControl extends React.Component {
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      config: this.props.config,
+      pullData: {gramsIn: 18, gramsOut: 36, coffee: "", pullTime: 30, flip: true}
+    };
+
+    this.handleChange = this.handleChange.bind(this);
+    this.handleSubmit = this.handleSubmit.bind(this);
+    this.handleConfigUpdate = this.handleConfigUpdate.bind(this);
+  }
+
+  componentDidUpdate(oldProps) {
+    if (oldProps !== this.props) {
+      let newState = {
+        config: {...this.props.config},
+        pullData: this.state.pullData,
+      }
+      this.setState(newState);
+    }
+  }
+
+  handleConfigUpdate(event) {
+    event.target.disabled = true;
+    let form = new FormData()
+    for (let [key, value] of Object.entries(this.state.config)) {
+      if (value == null) {
+        value = ""
+      }
+      form.append(key, value);
+    }
+    httpPostAsync("/config/", form, function(){
+      event.target.disabled = false;
+    })
+  }
+
+  handleChange(event) {
+    let val = event.target.value
+    if (event.target.type == "checkbox") {
+      val = event.target.checked
+    }
+    if (this.state.config.hasOwnProperty(event.target.name)) {
+      this.setState({
+        config: {
+          ...this.state.config,
+          [event.target.name]: val,
+        },
+      })
+    } else {
+      this.setState({
+        pullData: {
+          ...this.state.pullData,
+          [event.target.name]: val,
+        },
+      })
+    }
+  }
+
+  handleSubmit(event) {
+    event.target.disabled = true;
+    this.props.setConfig(this.state.config);
+    event.preventDefault();
+    captureImage(
+      {...this.state.config, ...this.state.pullData},
+      function(data) {event.target.disabled = false},
+    )
+  }
+
+  render() {
+    let configInputs = [];
+    for (let key in this.state.config) {
+      let value = this.state.config[key]
+      let labelName = key.replace(' ', '%nbsp;')
+      configInputs.push(
+        <div class="form-group">
+          <label>
+            {labelName}:
+            <input type="text" name={key} value={value} onChange={evt => this.handleChange(evt)} />
+          </label>
+        </div>
+      );
+    }
+    let captureInputs = [];
+    for (let key in this.state.pullData) {
```

Comment:
this is a fine way to loop and create jsx but there is a slightly sexier syntax available too...

Reply:
What is this sexy syntax I am unaware of?

Reply:
for example:
```<Labels>
          {_.map(labels, (label, index) => (
            <React.Fragment key={index}>
              {label}
            </React.Fragment>
          ))}
        </Labels>```

Reply:
Oh dear god...

Reply:
So that looks like it relies on Lodash?

Reply:
```
{this.state.config.map((value, key) => {
   const labelName = key.replace(' ', '%nbsp;')
return (
<div class="form-group">
          <label>
            {labelName}:
            <input type="text" name={key} value={value} onChange={evt => this.handleChange(evt)} />
          </label>
        </div>
);
})